tachyon/zk/air/plonky3/base/multi_field32_conversions.h

```diff
@@ -0,0 +1,62 @@
+// Copyright (c) 2022 The Plonky3 Authors
+// Use of this source code is governed by a MIT/Apache-2.0 style license that
+// can be found in the LICENSE-MIT.plonky3 and the LICENCE-APACHE.plonky3
+// file.
+
+#ifndef TACHYON_ZK_AIR_PLONKY3_BASE_MULTI_FIELD32_CONVERSIONS_H_
+#define TACHYON_ZK_AIR_PLONKY3_BASE_MULTI_FIELD32_CONVERSIONS_H_
+
+#include <stdint.h>
+
+#include <array>
+#include <limits>
+
+#include "absl/types/span.h"
+
+#include "tachyon/base/containers/adapters.h"
+#include "tachyon/build/build_config.h"
+
+namespace tachyon::zk::air::plonky3 {
+
+template <typename BigF, typename SmallF>
+BigF Reduce(absl::Span<const SmallF> values) {
+  static_assert(SmallF::Config::kModulusBits <= 32);
+  static_assert(BigF::Config::kModulusBits > 64);
+  const BigF kMultiplier(uint64_t{1} << 32);
+  BigF ret;
+  for (const SmallF& value : base::Reversed(values)) {
+    ret *= kMultiplier;
+    if constexpr (SmallF::Config::kUseMontgomery) {
+      ret += BigF(SmallF::Config::FromMontgomery(value.value()));
+    } else {
+      ret += BigF(value.value());
+    }
+  }
+  return ret;
+}
```

Comment:
Which is more efficient: accumulating smallF into BigF through addition or inserting smallF into each limb of BigF?

Reply:
I think the latter!